// @web5/crypto/dist/cjs/crypto-algorithms/ecdh.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.EcdhAlgorithm = void 0;
var utils_js_1 = require("../utils.js");
var index_js_1 = require("../crypto-primitives/index.js");
var index_js_2 = require("../algorithms-api/index.js");
var EcdhAlgorithm = /** @class */ (function (_super) {
    __extends(EcdhAlgorithm, _super);
    function EcdhAlgorithm() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.namedCurves = ['secp256k1', 'X25519'];
        return _this;
    }
    EcdhAlgorithm.prototype.deriveBits = function (options) {
        return __awaiter(this, void 0, void 0, function () {
            var algorithm, baseKey, length, sharedSecret, ownKeyAlgorithm, _a, ownPrivateKey, otherPartyPublicKey, ownPrivateKey, otherPartyPublicKey, lengthInBytes;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        algorithm = options.algorithm, baseKey = options.baseKey, length = options.length;
                        this.checkAlgorithmOptions({ algorithm: algorithm, baseKey: baseKey });
                        // The base key must be allowed to be used for deriveBits operations.
                        this.checkKeyUsages({ keyUsages: ['deriveBits'], allowedKeyUsages: baseKey.usages });
                        // The public key must be allowed to be used for deriveBits operations.
                        this.checkKeyUsages({ keyUsages: ['deriveBits'], allowedKeyUsages: algorithm.publicKey.usages });
                        ownKeyAlgorithm = baseKey.algorithm;
                        _a = ownKeyAlgorithm.namedCurve;
                        switch (_a) {
                            case 'secp256k1': return [3 /*break*/, 1];
                            case 'X25519': return [3 /*break*/, 3];
                        }
                        return [3 /*break*/, 5];
                    case 1:
                        ownPrivateKey = baseKey.material;
                        otherPartyPublicKey = algorithm.publicKey.material;
                        return [4 /*yield*/, index_js_1.Secp256k1.sharedSecret({
                                privateKey: ownPrivateKey,
                                publicKey: otherPartyPublicKey
                            })];
                    case 2:
                        sharedSecret = _b.sent();
                        return [3 /*break*/, 6];
                    case 3:
                        ownPrivateKey = baseKey.material;
                        otherPartyPublicKey = algorithm.publicKey.material;
                        return [4 /*yield*/, index_js_1.X25519.sharedSecret({
                                privateKey: ownPrivateKey,
                                publicKey: otherPartyPublicKey
                            })];
                    case 4:
                        sharedSecret = _b.sent();
                        return [3 /*break*/, 6];
                    case 5: throw new TypeError("Out of range: '".concat(ownKeyAlgorithm.namedCurve, "'. Must be one of '").concat(this.namedCurves.join(', '), "'"));
                    case 6:
                        // Length is null, return the full derived secret.
                        if (length === null)
                            return [2 /*return*/, sharedSecret];
                        // If the length is not a multiple of 8, throw.
                        if (length && length % 8 !== 0)
                            throw new index_js_2.OperationError("To be compatible with all browsers, 'length' must be a multiple of 8.");
                        lengthInBytes = length / 8;
                        // If length (converted to bytes) is larger than the derived secret, throw.
                        if (sharedSecret.byteLength < lengthInBytes)
                            throw new index_js_2.OperationError("Requested 'length' exceeds the byte length of the derived secret.");
                        // Otherwise, either return the secret or a truncated slice.
                        return [2 /*return*/, lengthInBytes === sharedSecret.byteLength ?
                                sharedSecret :
                                sharedSecret.slice(0, lengthInBytes)];
                }
            });
        });
    };
    EcdhAlgorithm.prototype.generateKey = function (options) {
        var _a;
        var _b;
        return __awaiter(this, void 0, void 0, function () {
            var algorithm, extractable, keyUsages, keyPair, cryptoKeyPair, _c;
            return __generator(this, function (_d) {
                switch (_d.label) {
                    case 0:
                        algorithm = options.algorithm, extractable = options.extractable, keyUsages = options.keyUsages;
                        this.checkGenerateKey({ algorithm: algorithm, keyUsages: keyUsages });
                        _c = algorithm.namedCurve;
                        switch (_c) {
                            case 'secp256k1': return [3 /*break*/, 1];
                            case 'X25519': return [3 /*break*/, 3];
                        }
                        return [3 /*break*/, 5];
                    case 1:
                        (_a = (_b = algorithm).compressedPublicKey) !== null && _a !== void 0 ? _a : (_b.compressedPublicKey = true);
                        return [4 /*yield*/, index_js_1.Secp256k1.generateKeyPair({
                                compressedPublicKey: algorithm.compressedPublicKey
                            })];
                    case 2:
                        keyPair = _d.sent();
                        return [3 /*break*/, 5];
                    case 3: return [4 /*yield*/, index_js_1.X25519.generateKeyPair()];
                    case 4:
                        keyPair = _d.sent();
                        return [3 /*break*/, 5];
                    case 5:
                        if (!(0, utils_js_1.isBytesKeyPair)(keyPair)) {
                            throw new Error('Operation failed to generate key pair.');
                        }
                        cryptoKeyPair = {
                            privateKey: new index_js_2.CryptoKey(algorithm, extractable, keyPair.privateKey, 'private', this.keyUsages.privateKey),
                            publicKey: new index_js_2.CryptoKey(algorithm, true, keyPair.publicKey, 'public', this.keyUsages.publicKey)
                        };
                        return [2 /*return*/, cryptoKeyPair];
                }
            });
        });
    };
    return EcdhAlgorithm;
}(index_js_2.BaseEcdhAlgorithm));
exports.EcdhAlgorithm = EcdhAlgorithm;
