// @web5/crypto/dist/cjs/crypto-algorithms/pbkdf2.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Pbkdf2Algorithm = void 0;
var index_js_1 = require("../algorithms-api/index.js");
var pbkdf2_js_1 = require("../crypto-primitives/pbkdf2.js");
var Pbkdf2Algorithm = /** @class */ (function (_super) {
    __extends(Pbkdf2Algorithm, _super);
    function Pbkdf2Algorithm() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.hashAlgorithms = ['SHA-256', 'SHA-384', 'SHA-512'];
        return _this;
    }
    Pbkdf2Algorithm.prototype.deriveBits = function (options) {
        return __awaiter(this, void 0, void 0, function () {
            var algorithm, baseKey, length, derivedBits;
            return __generator(this, function (_a) {
                algorithm = options.algorithm, baseKey = options.baseKey, length = options.length;
                this.checkAlgorithmOptions({ algorithm: algorithm, baseKey: baseKey });
                // The base key must be allowed to be used for deriveBits operations.
                this.checkKeyUsages({ keyUsages: ['deriveBits'], allowedKeyUsages: baseKey.usages });
                // If the length is 0, throw.
                if (typeof length !== 'undefined' && length === 0) {
                    throw new index_js_1.OperationError("The value of 'length' cannot be zero.");
                }
                // If the length is not a multiple of 8, throw.
                if (length && length % 8 !== 0) {
                    throw new index_js_1.OperationError("To be compatible with all browsers, 'length' must be a multiple of 8.");
                }
                derivedBits = pbkdf2_js_1.Pbkdf2.deriveKey({
                    hash: algorithm.hash,
                    iterations: algorithm.iterations,
                    length: length,
                    password: baseKey.material,
                    salt: algorithm.salt
                });
                return [2 /*return*/, derivedBits];
            });
        });
    };
    Pbkdf2Algorithm.prototype.importKey = function (options) {
        return __awaiter(this, void 0, void 0, function () {
            var format, keyData, algorithm, extractable, keyUsages, cryptoKey;
            return __generator(this, function (_a) {
                format = options.format, keyData = options.keyData, algorithm = options.algorithm, extractable = options.extractable, keyUsages = options.keyUsages;
                this.checkImportKey({ algorithm: algorithm, format: format, extractable: extractable, keyUsages: keyUsages });
                cryptoKey = new index_js_1.CryptoKey(algorithm, extractable, keyData, 'secret', keyUsages);
                return [2 /*return*/, cryptoKey];
            });
        });
    };
    return Pbkdf2Algorithm;
}(index_js_1.BasePbkdf2Algorithm));
exports.Pbkdf2Algorithm = Pbkdf2Algorithm;
