// @tbd54566975/dwn-sdk-js/dist/esm/src/interfaces/records-delete.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import { AbstractMessage } from '../core/abstract-message.js';
import { Message } from '../core/message.js';
import { Records } from '../utils/records.js';
import { Time } from '../utils/time.js';
import { DwnInterfaceName, DwnMethodName } from '../enums/dwn-interface-method.js';
export class RecordsDelete extends AbstractMessage {
    static parse(message) {
        return __awaiter(this, void 0, void 0, function* () {
            let signaturePayload;
            if (message.authorization !== undefined) {
                signaturePayload = yield Message.validateMessageSignatureIntegrity(message.authorization.signature, message.descriptor);
            }
            Records.validateDelegatedGrantReferentialIntegrity(message, signaturePayload);
            Time.validateTimestamp(message.descriptor.messageTimestamp);
            const recordsDelete = new RecordsDelete(message);
            return recordsDelete;
        });
    }
    /**
     * Creates a RecordsDelete message.
     * @param options.recordId If `undefined`, will be auto-filled as a originating message as convenience for developer.
     * @param options.messageTimestamp If `undefined`, it will be auto-filled with current time.
     */
    static create(options) {
        var _a;
        return __awaiter(this, void 0, void 0, function* () {
            const recordId = options.recordId;
            const currentTime = Time.getCurrentTimestamp();
            const descriptor = {
                interface: DwnInterfaceName.Records,
                method: DwnMethodName.Delete,
                recordId,
                messageTimestamp: (_a = options.messageTimestamp) !== null && _a !== void 0 ? _a : currentTime
            };
            const authorization = yield Message.createAuthorization({
                descriptor,
                signer: options.signer,
                protocolRole: options.protocolRole,
                delegatedGrant: options.delegatedGrant
            });
            const message = { descriptor, authorization };
            Message.validateJsonSchema(message);
            return new RecordsDelete(message);
        });
    }
}
