// @tbd54566975/dwn-sdk-js/dist/esm/src/interfaces/records-read.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import { getCurrentTimeInHighPrecision } from '../utils/time.js';
import { Message } from '../core/message.js';
import { ProtocolAuthorization } from '../core/protocol-authorization.js';
import { Records } from '../utils/records.js';
import { RecordsGrantAuthorization } from '../core/records-grant-authorization.js';
import { removeUndefinedProperties } from '../utils/object.js';
import { validateMessageSignatureIntegrity } from '../core/auth.js';
import { DwnInterfaceName, DwnMethodName } from '../core/message.js';
export class RecordsRead extends Message {
    static parse(message) {
        return __awaiter(this, void 0, void 0, function* () {
            if (message.authorization !== undefined) {
                yield validateMessageSignatureIntegrity(message.authorization.authorSignature, message.descriptor);
            }
            const recordsRead = new RecordsRead(message);
            return recordsRead;
        });
    }
    /**
     * Creates a RecordsRead message.
     * @param options.recordId If `undefined`, will be auto-filled as a originating message as convenience for developer.
     * @param options.date If `undefined`, it will be auto-filled with current time.
     *
     * @throws {DwnError} when a combination of required RecordsReadOptions are missing
     */
    static create(options) {
        var _a;
        return __awaiter(this, void 0, void 0, function* () {
            const { filter, authorizationSigner, permissionsGrantId, protocolRole } = options;
            const currentTime = getCurrentTimeInHighPrecision();
            const descriptor = {
                interface: DwnInterfaceName.Records,
                method: DwnMethodName.Read,
                filter: Records.normalizeFilter(filter),
                messageTimestamp: (_a = options.date) !== null && _a !== void 0 ? _a : currentTime,
            };
            removeUndefinedProperties(descriptor);
            // only generate the `authorization` property if signature input is given
            let authorization = undefined;
            if (authorizationSigner !== undefined) {
                authorization = yield Message.createAuthorizationAsAuthor(descriptor, authorizationSigner, { permissionsGrantId, protocolRole });
            }
            const message = { descriptor, authorization };
            Message.validateJsonSchema(message);
            return new RecordsRead(message);
        });
    }
    authorize(tenant, newestRecordsWrite, messageStore) {
        var _a;
        return __awaiter(this, void 0, void 0, function* () {
            const { descriptor } = newestRecordsWrite.message;
            // if author is the same as the target tenant, we can directly grant access
            if (this.author === tenant) {
                return;
            }
            else if (descriptor.published === true) {
                // authentication is not required for published data
                return;
            }
            else if (this.author !== undefined && this.author === descriptor.recipient) {
                // The recipient of a message may always read it
                return;
            }
            else if (descriptor.protocol !== undefined) {
                // All protocol RecordsWrites must go through protocol auth, because protocolPath, contextId, and record type must be validated
                yield ProtocolAuthorization.authorize(tenant, this, newestRecordsWrite, messageStore);
            }
            else if (this.author !== undefined && ((_a = this.authorSignaturePayload) === null || _a === void 0 ? void 0 : _a.permissionsGrantId) !== undefined) {
                yield RecordsGrantAuthorization.authorizeRead(tenant, this, newestRecordsWrite, this.author, messageStore);
            }
            else {
                throw new Error('message failed authorization');
            }
        });
    }
}
