// @web5/api/dist/esm/did-api.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import { DidMessage } from '@web5/agent';
/**
 * The DID API is used to resolve DIDs.
 *
 * @beta
 */
export class DidApi {
    constructor(options) {
        this.agent = options.agent;
        this.connectedDid = options.connectedDid;
    }
    /**
     * Resolves a DID to a DID Resolution Result.
     *
     * @param didUrl - The DID or DID URL to resolve.
     * @returns A promise that resolves to the DID Resolution Result.
     */
    resolve(didUrl, resolutionOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            const agentResponse = yield this.agent.processDidRequest({
                messageOptions: { didUrl, resolutionOptions },
                messageType: DidMessage.Resolve
            });
            const { result } = agentResponse;
            return result;
        });
    }
}
